### PR TITLE
104 Add login page

### PR DESCRIPTION
**Pull request by GeorgiChochov**
- Login/Logout are now restful.
- Simple login page inside the sapUI5 front-end, with routing that leads to the login page if you're unauthenticated and doesn't allow you to go back to it if you are.
- Supports multiple tabs being open, session lives on until it expires or user logs out.
- If user has two tabs open and logs out from one, when they try to do anything on the other they are sent to log in.

### CODE REVIEW

File: menas/src/main/scala/za/co/absa/enceladus/rest/WebSecurityConfig.scala [hunk 1]
```diff
@@ -116,24 +126,53 @@ class WebSecurityConfig {
...
+
+    @Autowired()
+    val authenticationFailureHandler: AuthenticationFailureHandler = null
+
+    @Autowired()
+    val logoutSuccessHandler: LogoutSuccessHandler = null
+
     override def configure(http: HttpSecurity) {
       http
+        .csrf().ignoringAntMatchers("/api/login").and()
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/rest/WebSecurityConfig.scala]: I would shift every `.XXX` onto new line, so it is the same as the rest

File: menas/ui/components/login/loginMain.controller.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      }
+    },
+
+    _resetFieldState: function (sField) {
+      this.byId(sField).setValueState(sap.ui.core.ValueState.None);
+      this.byId(sField).setValueStateText("");
+    },
+
+    _resetLoginFormState: function () {
+      this._resetFieldState("username");
```
Comment (Zejnilovic) [inline menas/ui/components/login/loginMain.controller.js]: Here comes the question/discussion. Do we want to be the guys that make you re-type even the username?

File: menas/ui/components/login/loginMain.controller.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        },
+        data: oData,
+        method: "POST",
+        success: fnSuccess,
+        error: fnError
+      })
+    },
+    /**
+     * Similar to onAfterRendering, but this hook is invoked before the controller's
+     * View is re-rendered (NOT before the first rendering! onInit() is used for
```
Comment (Zejnilovic) [inline menas/ui/components/login/loginMain.controller.js]: Is this left here on purpose?

File: menas/ui/generic/functions.js [hunk 4]
```diff
@@ -13,34 +13,42 @@
...
-			method : sMethod,
-			success : fnSuccess,
-			error : fnError
-		})
-	}
+  this.ajax = function (sPath, sMethod, oData, fnSuccess, fnError, oControl) {
+    if (oControl) oControl.setBusy(true);
+
+    let oFormattedData = null;
+    if ((sMethod.toLowerCase() === "post" || sMethod.toLowerCase() === "put") && typeof (oData) === "object") {
```
Comment (Zejnilovic) [inline menas/ui/generic/functions.js]: maybe a new line after `&&` or something like that? And put the whole last type of in `()`

Comment (Zejnilovic) [reply menas/ui/generic/functions.js]: Is there a reason for `oData` to be in `()`

File: menas/ui/components/login/loginMain.view.xml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+<core:View id="loginMainView" xmlns:core="sap.ui.core" xmlns:mvc="sap.ui.core.mvc" xmlns="sap.m" xmlns:form="sap.ui.layout.form"
+		   controllerName="components.login.loginMain">
+	<Page title="Login" enableScrolling="true">
+		<content>
+			<VBox id="loginForm" fitContainer="true" justifyContent="Center" alignItems="Center">
+				<items>
+					<Input type="Text" id="username" placeholder="username" value="{login>/username}" submit="onLoginSubmit"/>
+					<Input type="Password" id="password" placeholder="password" value="{login>/password}" submit="onLoginSubmit"/>
+					<Button width="10rem" text="Login" type="Emphasized" id="loginSubmit"/>
```
Comment (Zejnilovic) [inline menas/ui/components/login/loginMain.view.xml]: I know there is no predetermined sequence of attributes, but I would like if we could keep it same for the whole view. So `type`, `id`, `..rest..`. Just for readability.

File: menas/ui/components/login/loginMain.controller.js (hunk 3 above)
Comment (GeorgiChochov) [reply menas/ui/components/login/loginMain.controller.js]: Yeah, for consistency like in the other controllers. Maybe we can drop them, I don't know.

File: menas/ui/components/login/loginMain.view.xml (hunk 5 above)
Comment (GeorgiChochov) [reply menas/ui/components/login/loginMain.view.xml]: Fair point.

File: menas/ui/generic/functions.js (hunk 4 above)
Comment (GeorgiChochov) [reply menas/ui/generic/functions.js]: > Is there a reason for `oData` to be in `()`

Don't think so, just the way it was.

File: menas/ui/components/login/loginMain.controller.js (hunk 2 above)
Comment (GeorgiChochov) [reply menas/ui/components/login/loginMain.controller.js]: You saying extract it? I can do that.

File: menas/ui/generic/functions.js [hunk 6]
```diff
@@ -13,34 +13,45 @@
...
-			success : fnSuccess,
-			error : fnError
-		})
-	}
+  this.ajax = function (sPath, sMethod, oData, fnSuccess, fnError, oControl) {
+    if (oControl) oControl.setBusy(true);
+
+    let oFormattedData = null;
+    if ((sMethod.toLowerCase() === "post" || sMethod.toLowerCase() === "put")
+      && (typeof oData === "object")) {
```
Comment (Zejnilovic) [inline menas/ui/generic/functions.js]: I think `&&` should be the last thing on the previous line. So the programmer reading it know that it is continuing from a first glance.

Comment (GeorgiChochov) [reply menas/ui/generic/functions.js]: I tend to disagree, for the same reason method calls on new lines start with a dot (e.g. `.map(...)`) makes it clear that this line isn't a statement on its own. I look at the start of lines a lot more than their ends of I'm just scanning the code.

Comment (Zejnilovic) [reply menas/ui/generic/functions.js]: As this is subjective, dealer call. Just know, that people who look at the end exist 😄

**Pull request merged**